compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -0,0 +1,951 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import core._
+import Contexts._
+import Symbols._
+import Types._
+import StdNames._
+
+import ast.tpd._
+import util.EqHashMap
+import config.Printers.init as printer
+import reporting.trace as log
+
+import Errors._
+
+import scala.collection.mutable
+
+class Semantic {
+  import Semantic._
+
+// ----- Domain definitions --------------------------------
+
+  /** Abstract values
+   *
+   *  Value = Hot | Cold | Warm | ThisRef | Fun | RefSet
+   *
+   *                 Cold
+   *        ┌──────►  ▲   ◄──┐  ◄────┐
+   *        │         │      │       │
+   *        │         │      │       │
+   *   ThisRef(C)     │      │       │
+   *        ▲         │      │       │
+   *        │     Warm(D)   Fun    RefSet
+   *        │         ▲      ▲       ▲
+   *        │         │      │       │
+   *      Warm(C)     │      │       │
+   *        ▲         │      │       │
+   *        │         │      │       │
+   *        └─────────┴──────┴───────┘
+   *                  Hot
+   */
```

Comment:
I like the diagram but I think RefSet should be above ThisRef(C), Warm(D), and Fun (but below Cold).

Reply:
Putting `RefSet` above `ThisRef(C)` is also problematic (it might not be a member). In the end, I added more text to explain the ordering with `RefSet`.